fdbserver/CommitProxyServer.actor.cpp

```diff
@@ -874,6 +877,81 @@ ACTOR Future<Void> preresolutionProcessing(CommitBatchContext* self) {
 	return Void();
 }
 
+namespace {
+// Routine allows caller to find TenantName for a given TenantId. It returns empty optional when either TenantId is
+// invalid or tenant is unknown
+Optional<TenantName> getTenantName(ProxyCommitData* commitData, int64_t tenantId) {
+	if (tenantId != TenantInfo::INVALID_TENANT) {
+		auto itr = commitData->tenantIdIndex.find(tenantId);
+		if (itr != commitData->tenantIdIndex.end()) {
+			return Optional<TenantName>(itr->second);
+		}
+	}
+
+	return Optional<TenantName>();
+}
+
+std::pair<EncryptCipherDomainName, EncryptCipherDomainId> getEncryptDetailsFromMutationRef(ProxyCommitData* commitData,
+                                                                                           MutationRef m) {
+	std::pair<EncryptCipherDomainName, EncryptCipherDomainId> details(EncryptCipherDomainName(),
+	                                                                  INVALID_ENCRYPT_DOMAIN_ID);
+
+	// Possible scenarios:
+	// 1. Encryption domain (Tenant details) weren't explicitly provided, extract Tenant details using
+	// TenantPrefix (first 8 bytes of FDBKey)
+	// 2. Encryption domain isn't available, leverage 'default encryption domain'
+
+	if (commitData->tenantMap.empty()) {
+		// Cluster serves no-tenants; use 'default encryption domain'
```

Comment:
I guess we still need to distinguish here between system and normal keys, is that right? You could probably cover this like:

```
if (isSystemKey(m.param1)) { ... } // works even for clear range
else if (commitData->tenantMap.empty()) { ... }
else if (isSingleKeyMutation(...)) { ... }
else { ... } // Clear range that starts in the normal keys, though may still terminate in system keys
```

Reply:
Good catch! We can always encrypt systemKeys with its domain regardless of TenantMap.